MAINT: SCIPY_USE_PROPACK

tylerjereddy opened the pull request:
* this is effectively a forward port and modernization
of the release branch `PROPACK` shims that were added in `1.8.x` (but not `main`)
gh-15432; 
in short, `PROPACK` + Windows + some linalg backends
was causing all sorts of trouble, and this has never been resolved

* I've switched to `SCIPY_USE_PROPACK` instead of `USE_PROPACK`
for the opt-in, since this was requested by @h-vetinari, though the change
between release branches may cause a little confusion (another
release note adjustment to add maybe)

* I think the issues are painful to reproduce; for my part,
I did the following just to check the proper skipping/accounting
of tests:

- `SCIPY_USE_PROPACK=1 python dev.py -j 20 -t scipy/sparse/linalg`
  - `932 passed, 172 skipped, 8 xfailed in 115.57s (0:01:55)`
- `python dev.py -j 20 -t scipy/sparse/linalg`
  - `787 passed, 317 skipped, 8 xfailed in 114.80s (0:01:54)`

* why am I doing this now? well, to be frank the process of manually
backporting this for each release is error-prone, and may cause
additional confusion/debate, which I'd like to avoid. Besides, if it
is broken in `main` we may as well have the shims there as well. I would
point out that you may want to add `SCIPY_USE_PROPACK` to 1 or 2 jobs
in CI? The other reason is that if usage of `PROPACK` spreads, I don't
want to be manually applying more skips/shims on each release (which
I already had to do here with two new tests it seems)

* once finalized, I would plan to backport this to `maintenance/1.9.x`

h-vetinari commented:
> I've switched to SCIPY_USE_PROPACK instead of USE_PROPACK for the opt-in, since this was requested by @h-vetinari, ...

Just to note that this was more of a question/suggestion than a request. :)

h-vetinari commented on scipy/sparse/linalg/_eigen/_svds.py:
```diff
@@ -297,6 +302,10 @@ def matmat_XH_X(x):
         eigvec, _ = np.linalg.qr(eigvec)
 
     elif solver == 'propack':
+        if not HAS_PROPACK:
+            raise ValueError("`solver='propack'` is opt-in due to potential issues on Windows, "
+                             "it can be enabled by setting the `USE_PROPACK` environment "
+                             "variable before importing scipy")
```
Needs an update of the environment variable here too.

tylerjereddy replied:
Thanks, I've pushed that adjustment in. So far, I've decided to ignore the linter complaints though, unless there are strong objections to that.

WarrenWeckesser replied:
Please fix the line length.  There is no reason to not remain in compliance with such a simple formatting rule.

tylerjereddy commented:
@mckib2 shall we add `SCIPY_USE_PROPACK` to at least one CI job then?

mckib2 commented:
> @mckib2 shall we add `SCIPY_USE_PROPACK` to at least one CI job then?

Yes!  If it's not already running in the GitHub Meson Linux and MacOS pipelines, that might be a good place for them.  Did you want a Windows build as well or will we poke that beast when those issues are fully ironed out?

tylerjereddy commented:
Ok, I added `SCIPY_USE_PROPACK=1` to `meson` testing in CI on all 3 platforms and fixed the linting issues. `PROPACK` was being used everywhere before so I see no reason not to test on all 3 platforms now--the real problem is when you swap linalg backends on `conda-forge`, etc. I think (and maybe with 32-bit Windows sometimes, based on wheels repo...).

If there are any other nitpicks feel free to push them right in. I see that the patch diff coverage may suggest that the `meson` jobs are not reporting coverage to `codecov` -- I have no comment on that other than pointing this out to reviewers who may be curious about the numbers there.

rgommers reviewed (approved):
Thanks @tylerjereddy & all. This looks like a good idea to me. Note that the issue for tracking the PROPACK segfault is gh-15108.

CI is green, and this should be good to backport to 1.9.x as is I think.

rgommers commented:
> I see that the patch diff coverage may suggest that the `meson` jobs are not reporting coverage to `codecov` -- I have no comment on that other than pointing this out to reviewers who may be curious about the numbers there.

There's something off with `gcov` integration, see gh-15896. I'm aware of it, just not highest prio right now.
